<div style="text-align: center; padding: 20px; font-family: Vazir;">
<h1 align="center" style="font-size: 28px; color:rgb(64, 244, 202); width: 100%;">⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️<br>تمرین 4<br>⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️</h1>
<h2 style="color:rgb(90, 255, 184); font-size: 20px;">Comparing BART and GPT-2 for Text-to-SQL</h2>
<p align="center" style="color: #666; font-size: 16px;">زهرا فیض اله زاده</p>
<p align="center" style="color: #666; font-size: 16px; margin-bottom: 30px;">zahra.feizolahzadeh@gmail.com</p>

<div dir="rtl" style="border: 2px dashed rgb(90, 255, 184); border-radius: 8px; padding: 20px; margin: 20px auto; max-width: 500px; text-align: right;">
<p style="color: rgb(64, 244, 202); font-size: 18px; margin-bottom: 15px;">📝 مشخصات دانشجو:</p>
<p style="color: #666; margin: 5px;">نام و نام خانوادگی: امین آقاکثیری</p>
<p style="color: #666; margin: 5px;">شماره دانشجویی: 810101381</p>
<p style="color: #666; margin: 5px;">تاریخ ارسال:1404/9/26 </p>
</div>
</div>

<div dir="rtl" style="text-align: justify; padding: 25px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
<div style="line-height: 2.0; font-size: 17px; color: black; font-family: Vazir;">
<div style="padding-right:40px">
</div>
<br>
<div style="padding-right:100px">
📋 <b>ساختار تمرین:</b>
<br>
<li><b>سوال اول - <span dir="ltr">Text-to-SQL with BART and GPT-2</span> (50)</b></li>
<ul>
<li>بخش اول: بارگذاری و پیش‌پردازش داده‌های Text-to-SQL</li>
<li>بخش دوم: آموزش مدل Encoder–Decoder (BART)</li>
<li>بخش سوم: آموزش مدل Decoder-only (GPT-2)</li>
<li>بخش چهارم: مقایسه و تحلیل نهایی</li>

</ul>

</div>
</div>
<div dir='rtl' style="line-height: 1.8; font-family: Vazir; font-size: 16px; margin-top: 20px; background-color: #e8eaf6; padding: 15px; border-radius: 8px; color:black">
💡 <b>نکات مهم:</b>
<br>
<li>	مسئله Text-to-SQL حساسیت بسیار بالایی به جزئیات دارد.</li>
کوچک‌ترین اختلاف در ساختار SQL، ترتیب کلیدواژه‌ها یا حتی فاصله‌ها می‌تواند باعث نامعتبر شدن خروجی شود. بنابراین هنگام ارزیابی، به‌ویژه در معیار Normalized EM، باید دقت زیادی به اصلاح فاصله‌ها، حذف سمیکالن اضافه، و یکسان‌سازی case کلمات داشته باشید تا مقایسه‌ای منصفانه انجام شود.
<li>	معماری‌های Encoder–Decoder و Decoder-only رفتارهای متفاوتی در تولید SQL دارند.</li>
ساخت صحیح ورودی‌ها (Prefix)، نحوه قرار دادن question و schema در متن، ترتیب tokenها،          padding مناسب و mask کردن بخش‌های غیرهدف، همگی نقش اساسی در عملکرد این مدل‌ها دارند. دقت کنید که GPT-2 باید تمام اطلاعات را در قالب یک توالی یکپارچه دریافت کند، درحالی‌که BART ورودی را رمزگذاری و سپس خروجی SQL را مستقل تولید می‌کند.
<li>	برای مقایسه مدل‌ها، آموزش حداقلی اما واقعی لازم است.</li>
باید حداقل یک ایپاک کامل روی حداقل بیست هزار نمونه از داده‌های آموزش اجرا شود تا مدل به‌طور معنی‌دار الگوهای اولیه Text-to-SQL را یاد بگیرد و ارزیابی‌ها قابل اعتماد باشند. لازم نیست آموزش طولانی انجام دهید، اما اجرای همین مقدار حداقلی برای تحلیل و مقایسه معماری‌ها ضروری است.
</div>
</div>

# <div style="text-align: center; direction: rtl; font-family: Vazir;"><h1 align="center" style="font-size: 24px; padding: 20px;">⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️<br>سوال دوم: تحلیل و مقایسه معماری‌های Encoder–Decoder و Decoder-Only در تسک Text-to-SQL <br>⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️</h1></div>


<p dir="rtl" style="text-align: right; padding:30px; background-color:rgb(12, 12, 12); border-radius: 12px; color: white; font-family: Vazir;">
در این بخش از تمرین با مسئله Text-to-SQL آشنا می‌شوید؛ وظیفه‌ای که در آن مدل باید پرسش‌های زبان طبیعی را به کوئری SQL معتبر تبدیل کند. هدف تمرین، مقایسه عملی دو معماری مهم ترنسفورمری است: Encoder–Decoder‌ (مانند BART) و Decoder-only‌ (مانند GPT-2). پس از آماده‌سازی داده‌ها، ابتدا مدل BART و سپس مدل GPT-2 را برای این تسک آموزش می‌دهید و در نهایت عملکرد دو مدل را از نظر دقت، رفتار تولیدی و انواع خطاها تحلیل و مقایسه خواهید کرد.
</p>




## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش اول: مجموعه‌داده و پیش‌پردازش(10 نمره)</div>


<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
داده‌های Gretel Synthetic Text-to-SQL را از HuggingFace بارگذاری کرده و فایل‌های train/test را دریافت کنید. ساختار دیتاست را بررسی کرده، یک نمونه نمایش دهید و نقش ستون‌های اصلی از جمله question، schema و query را توضیح دهید. سپس داده‌ها را با استفاده از train_test_split به بخش‌های train و dev تقسیم کنید و ستون‌ها را به اسامی استاندارد (question، schema، query) تغییر دهید. برای کاهش زمان آموزش، روی هر بخش محدودیت اندازه تعیین کنید. در پایان، تحلیل کوتاهی از ماهیت پرسش‌ها، ساختار schemaها و چالش‌هایی که برای مدل ایجاد می‌کنند ارائه دهید.
</div>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
در پایان این بخش انتظار می‌رود دانشجو یک دیتافریم کاملا آماده برای مراحل بعدی در اختیار داشته باشد. خروجی مطلوب شامل بارگذاری صحیح فایل‌های train و test، نمایش ساختار کلی داده‌ها، تقسیم درست داده‌ها به بخش‌های train و dev، تغییر نام ستون‌ها به قالب استاندارد، و اعمال زیرنمونه‌گیری برای محدود کردن حجم داده‌ها است. علاوه بر این، باید یک نمونه واقعی از داده‌ها همراه با توضیح مختصر درباره ماهیت سؤال، ساختار schema و نمونه SQL نمایش داده شود تا مشخص شود داده‌ها به‌درستی پردازش شده و برای ورود به مراحل آموزش مدل آماده هستند.

</p>

In [1]:
%%capture
!pip install -U protobuf==3.20.3 transformers datasets sqlparse

In [2]:
from datasets import load_dataset
import pandas as pd
import re
from sklearn.model_selection import train_test_split
import torch
from torch.utils.data import Dataset, DataLoader
from torch.optim import AdamW
from transformers import BartTokenizer, BartForConditionalGeneration
from transformers import GPT2Tokenizer, GPT2LMHeadModel
from transformers import get_linear_schedule_with_warmup
import time
import sqlparse
from tqdm import tqdm
import random

In [3]:
dataset = load_dataset("gretelai/synthetic_text_to_sql")

README.md: 0.00B [00:00, ?B/s]

synthetic_text_to_sql_train.snappy.parqu(…):   0%|          | 0.00/32.4M [00:00<?, ?B/s]

synthetic_text_to_sql_test.snappy.parque(…):   0%|          | 0.00/1.90M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/100000 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/5851 [00:00<?, ? examples/s]

In [ ]:
print(dataset)

DatasetDict({
    train: Dataset({
        features: ['id', 'domain', 'domain_description', 'sql_complexity', 'sql_complexity_description', 'sql_task_type', 'sql_task_type_description', 'sql_prompt', 'sql_context', 'sql', 'sql_explanation'],
        num_rows: 100000
    })
    test: Dataset({
        features: ['id', 'domain', 'domain_description', 'sql_complexity', 'sql_complexity_description', 'sql_task_type', 'sql_task_type_description', 'sql_prompt', 'sql_context', 'sql', 'sql_explanation'],
        num_rows: 5851
    })
})


In [4]:
train_df = pd.DataFrame(dataset['train'])
test_df = pd.DataFrame(dataset['test'])

In [ ]:
train_df.head()

id              domain  \
0  5097            forestry   
1  5098    defense industry   
2  5099      marine biology   
3  5100  financial services   
4  5101              energy   

                                  domain_description    sql_complexity  \
0  Comprehensive data on sustainable forest manag...       single join   
1  Defense contract data, military equipment main...       aggregation   
2  Comprehensive data on marine species, oceanogr...         basic SQL   
3  Detailed financial data including investment s...       aggregation   
4  Energy market data covering renewable energy s...  window functions   

                          sql_complexity_description            sql_task_type  \
0        only one join (specify inner, outer, cross)  analytics and reporting   
1  aggregation functions (COUNT, SUM, AVG, MIN, M...  analytics and reporting   
2           basic SQL with a simple select statement  analytics and reporting   
3  aggregation functions (COUNT, SUM, AVG, MIN, M...  analytics and reporting   
4  window functions (e.g., ROW_NUMBER, LEAD, LAG,...  analytics and reporting   

                           sql_task_type_description  \
0  generating reports, dashboards, and analytical...   
1  generating reports, dashboards, and analytical...   
2  generating reports, dashboards, and analytical...   
3  generating reports, dashboards, and analytical...   
4  generating reports, dashboards, and analytical...   

                                          sql_prompt  \
0  What is the total volume of timber sold by eac...   
1  List all the unique equipment types and their ...   
2  How many marine species are found in the South...   
3  What is the total trade value and average pric...   
4  Find the energy efficiency upgrades with the h...   

                                         sql_context  \
0  CREATE TABLE salesperson (salesperson_id INT, ...   
1  CREATE TABLE equipment_maintenance (equipment_...   
2  CREATE TABLE marine_species (name VARCHAR(50),...   
3  CREATE TABLE trade_history (id INT, trader_id ...   
4  CREATE TABLE upgrades (id INT, cost FLOAT, typ...   

                                                 sql  \
0  SELECT salesperson_id, name, SUM(volume) as to...   
1  SELECT equipment_type, SUM(maintenance_frequen...   
2  SELECT COUNT(*) FROM marine_species WHERE loca...   
3  SELECT trader_id, stock, SUM(price * quantity)...   
4  SELECT type, cost FROM (SELECT type, cost, ROW...   

                                     sql_explanation  
0  Joins timber_sales and salesperson tables, gro...  
1  This query groups the equipment_maintenance ta...  
2  This query counts the number of marine species...  
3  This query calculates the total trade value an...  
4  The SQL query uses the ROW_NUMBER function to ...

In [5]:
# change col names
mapping = {
    'sql_prompt': 'question',
    'sql_context': 'schema',
    'sql': 'query'
}

train_df = train_df.rename(columns=mapping)
test_df = test_df.rename(columns=mapping)

train_df = train_df[['question', 'schema', 'query']]
test_df = test_df[['question', 'schema', 'query']]

print(f"New columns: {train_df.columns.tolist()}")


New columns: ['question', 'schema', 'query']


In [6]:
# train/dev split
train_data, dev_data = train_test_split(
    train_df,
    test_size=0.1,
    random_state=30
)

print(f"Numbers after split:")
print(f"Train: {len(train_data)}")
print(f"Dev: {len(dev_data)}")
print(f"Test: {len(test_df)}")

Numbers after split:
Train: 90000
Dev: 10000
Test: 5851


In [7]:
# train with 20000 train samples, 500 dev samples and 2000 test samples (as the question said 20000 is ok for training)
train_df = train_data.sample(n=20000, random_state=30).reset_index(drop=True)
dev_df = dev_data.sample(n=min(500, len(dev_data)), random_state=30).reset_index(drop=True)
test_df = test_df.sample(n=2000, random_state=30).reset_index(drop=True)

print("NUmbers after subsampling:")
print(f"Train: {len(train_df)}")
print(f"Dev: {len(dev_df)}")
print(f"Test: {len(test_df)}")

NUmbers after subsampling:
Train: 20000
Dev: 500
Test: 2000


In [ ]:
sample = train_df.iloc[10]

print("Sample question:\n")
print(sample['question'] + "\n")
print(50*"=")

print("Sample schema:\n")
print(sample['schema'] + "\n")
print(50*"=")

print("Sample query:\n")
print(sample['query'])

Sample question:

What is the local economic impact of sustainable tourism in Rome?

Sample schema:

CREATE TABLE local_impact (city TEXT, sustainability_score INT, economic_impact INT); INSERT INTO local_impact (city, sustainability_score, economic_impact) VALUES ('Rome', 8, 5000000), ('Rome', 9, 6000000);

Sample query:

SELECT economic_impact FROM local_impact WHERE city = 'Rome';


<div dir="rtl" style="background:#fffbe6;font-family: Vazir;border:1px dashed #f0ad4e;padding:12px;border-radius:8px;color:#111;line-height:1.9;overflow:auto;max-width:100%;box-sizing:border-box;">
✍️ <b>پاسخ تشریحی:</b><br>

ابتدا کمی درباره ۳ ستون توضیح می دهم. ستون question حاوی پرسش کاربر به زبان طبیعی است. وظیفه مدل این است که این متن را درک کند و بفهمد کاربر دقیقا چه اطلاعاتی می‌ خواهد.
ستون schema ساختار پایگاه داده را مشخص می‌ کند. این ستون شامل دستورات CREATE TABLE است که نشان می‌ دهد چه جداولی وجود دارند، هر جدول چه ستون‌هایی دارد، نوع داده هر ستون چیست و روابط بین جداول چگونه تعریف شده. مدل بدون این اطلاعات نمی‌تواند بداند از چه نام‌هایی برای جداول و ستون‌ها استفاده کند.
ستون query خروجی مورد انتظار است. یعنی کوئری SQL صحیحی که پاسخ سوال کاربر را از روی Schema داده شده استخراج می‌ کند. این ستون در واقع label آموزشی ماست که مدل باید یاد بگیرد آن را تولید کند.

در این دیتاست، برخی سوالات ساده هستند و فقط به یک جدول مراجعه می‌کنند ولی برخی دیگر پیچیده‌ تر اند و نیاز به ترکیب چند جدول یا استفاده از توابعی مانند JOIN دارند. نکته مهم این است که یک مفهوم واحد می‌ تواند به چندین شکل مختلف پرسیده شود که این موضوع یادگیری الگو های کلی را برای مدل دشوار می‌ کند.

هر نمونه شامل یک Schema است که معمولا با دستور CREATE TABLE تعریف شده. این Schema‌ ها اطلاعاتی مانند نام جداول، نام ستون‌ ها، نوع داده هر ستون و روابط بین جداول را در خود دارند. طول Schema‌ ها بسیار متفاوت است به طوری که برخی فقط یک جدول ساده دارند و برخی دیگر شامل چندین جدول با روابط پیچیده شامل Foreign Key هستند. این تنوع در طول ورودی می‌تواند مشکلاتی در tokenization و محدودیت طول context ایجاد کند.

حال درباره چالش های این مدل سازی، اولین و مهم‌ ترین چالش، حساسیت بالای SQL به جزئیات است. یک کاما یا پرانتز جا افتاده می‌تواند کل کوئری را نامعتبر کند. این یعنی مدل باید بتواند ساختار دستوری SQL را به طور کامل یاد بگیرد.

چالش دوم مربوط به نگاشت بین زبان طبیعی و SQL است. گاهی یک سوال می‌تواند چندین پاسخ SQL معتبر داشته باشد که همگی نتیجه یکسانی برمی‌ گردانند. این موضوع ارزیابی مدل را پیچیده می‌ کند چون Exact Match ممکن است پاسخ‌های صحیح اما متفاوت را رد کند.

چالش سوم تفاوت در نحوه پردازش ورودی توسط دو معماری است. در BART ورودی (question و Schema) ابتدا کامل encode می‌شود و سپس SQL تولید می‌شود، در حالی که GPT-2 باید همه چیز را در یک دنباله واحد ببیند و به صورت چپ به راست پردازش کند. این تفاوت بنیادی بر نحوه طراحی ورودی و همچنین عملکرد نهایی مدل‌ها تاثیر مستقیم دارد.

</div>


## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش دوم: آموزش مدل Encoder–Decoder (BART) (۱۵ نمره)</div>


<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
<p style="line-height: 1.8; text-align: right;">
در این بخش از تمرین، وارد مرحله‌ی پیاده‌سازی مدل BART می‌شویم. پیش از شروع، لازم است به‌صورت مختصر و دقیق توضیح دهید که مدل BART چیست و چگونه کار می‌کند؛ به‌ویژه این‌که چرا معماری Encoder–Decoder آن برای وظایفی مانند Text-to-SQL مناسب است. انتظار می‌رود دانشجو به ساختار دو‌بخشی BART، نحوهٔ پردازش ورودی در انکودر و تولید خروجی در دیکودر اشاره کند و نشان دهد که این معماری برای نگاشت دنباله‌ای از کلمات (سؤال + schema) به دنباله‌ای دیگر (SQL) کاملاً مناسب است.
پس از توضیح مدل، باید قالب ورودی مناسبی طراحی کنید که در آن سؤال و schema به‌صورت یک رشتهٔ یکپارچه در اختیار مدل قرار گیرند. سپس یک Dataset اختصاصی برای BART بسازید که وظایفی مانند tokenize کردن ورودی، ساخت attention mask‌ها و آماده‌سازی SQL به عنوان خروجی هدف مدل را انجام دهد. مدل را <strong>حداقل یک ایپاک روی بیست هزار نمونه</strong> آموزش دهید و loss نهایی، زمان اجرا و هایپرپارامترهای اصلی را گزارش کنید.
در مرحلهٔ ارزیابی، باید عملکرد مدل را روی dev و test با دو معیار Raw Exact Match و Normalized Exact Match بسنجید.
از شما انتظار می‌رود که هر دو معیار را توضیح دهید:
Raw EM چگونه مقایسهٔ مستقیم دو رشته را انجام می‌دهد، چرا ممکن است برای SQL ناکافی باشد و Normalized EM چگونه با حذف تفاوت‌های سطحی (مثل فاصله‌ها، علامت‌ها و بزرگی یا کوچکی حروف) مقایسه‌ای عادلانه‌تر ارائه می‌دهد.
در پایان، پنج نمونهٔ تصادفی از dev انتخاب کنید و پرسش، خروجی مدل، SQL طلایی و نوع خطاها را تحلیل کنید.
<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
نکته مهم:
برای پیاده‌سازی Raw EM می‌توانید از مقایسهٔ مستقیم رشته‌ها در پایتون استفاده کنید.
برای Normalized EM مجاز هستید از ابزارهایی مانند sqlparse برای فرمت‌بندی SQL و توابع سادهٔ پردازش متن جهت lowercase کردن و حذف فاصله‌های اضافی استفاده کنید، اما منطق مقایسه‌ی Normalized باید توسط خودتان نوشته شود.
</p>
</div>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
خروجی مورد انتظار این بخش، یک مدل BART آموزش‌دیده و ارزیابی‌شده است که بتواند از روی پرسش و ساختار پایگاه‌داده، کوئری SQL معتبر تولید کرده و معیارهای Raw EM و Normalized EM آن روی داده‌های dev و test گزارش شده باشد.
</p>

In [41]:
class BARTDataset(Dataset):
    def __init__(self, df, tokenizer, max_inp_tok=1024, max_out_tok=256):
        self.data = df.reset_index(drop=True)
        self.tokenizer = tokenizer
        self.max_inp_tok = max_inp_tok
        self.max_out_tok = max_out_tok

    def __len__(self):
        return len(self.data)

    def __getitem__(self, index):
        row = self.data.iloc[index]

        inp_text = (
            "Task: Translate given natural language Question about given Schema into SQL.\n"
            f"Question: {row['question']}\n"
            f"Schema: {row['schema']}"
        )
        target_text = row['query']

        inp_encoding = self.tokenizer(
            inp_text,
            max_length=self.max_inp_tok,
            padding='max_length',
            truncation=True,
            return_tensors='pt'
        )

        target_encoding = self.tokenizer(
            target_text,
            max_length=self.max_out_tok,
            padding='max_length',
            truncation=True,
            return_tensors='pt'
        )

        labels = target_encoding['input_ids'].squeeze()
        # if it was a pad token, we should mask it (mask value for torch is -100)
        # in order to not consider these tokens in learning process
        labels[labels == self.tokenizer.pad_token_id] = -100

        return {
            'input_ids': inp_encoding['input_ids'].squeeze(),
            'attention_mask': inp_encoding['attention_mask'].squeeze(),
            'labels': labels
        }

In [42]:
tokenizer = BartTokenizer.from_pretrained('facebook/bart-base')
model = BartForConditionalGeneration.from_pretrained('facebook/bart-base')
print("BART tokenizer and model loaded")

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

config.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/558M [00:00<?, ?B/s]

BART tokenizer and model loaded


In [43]:
print(f"Number of parameters: {model.num_parameters()}")
print(model.config)

Number of parameters: 139420416
BartConfig {
  "activation_dropout": 0.1,
  "activation_function": "gelu",
  "add_bias_logits": false,
  "add_final_layer_norm": false,
  "architectures": [
    "BartModel"
  ],
  "attention_dropout": 0.1,
  "bos_token_id": 0,
  "classif_dropout": 0.1,
  "classifier_dropout": 0.0,
  "d_model": 768,
  "decoder_attention_heads": 12,
  "decoder_ffn_dim": 3072,
  "decoder_layerdrop": 0.0,
  "decoder_layers": 6,
  "decoder_start_token_id": 2,
  "dropout": 0.1,
  "dtype": "float32",
  "early_stopping": true,
  "encoder_attention_heads": 12,
  "encoder_ffn_dim": 3072,
  "encoder_layerdrop": 0.0,
  "encoder_layers": 6,
  "eos_token_id": 2,
  "forced_bos_token_id": 0,
  "forced_eos_token_id": 2,
  "gradient_checkpointing": false,
  "id2label": {
    "0": "LABEL_0",
    "1": "LABEL_1",
    "2": "LABEL_2"
  },
  "init_std": 0.02,
  "is_encoder_decoder": true,
  "label2id": {
    "LABEL_0": 0,
    "LABEL_1": 1,
    "LABEL_2": 2
  },
  "max_position_embeddings": 1024

In [44]:
train_dataset = BARTDataset(train_df, tokenizer)
dev_dataset = BARTDataset(dev_df, tokenizer)
test_dataset = BARTDataset(test_df, tokenizer)

print(f"Train dataset size: {len(train_dataset)}")
print(f"Dev dataset size: {len(dev_dataset)}")
print(f"Test dataset size: {len(test_dataset)}")

Train dataset size: 20000
Dev dataset size: 500
Test dataset size: 2000


In [45]:
BATCH_SIZE = 16
EPOCHS = 1
LEARNING_RATE = 6e-5

In [46]:
# making data loaders
train_loader = DataLoader(train_dataset, batch_size=BATCH_SIZE, shuffle=True)
dev_loader = DataLoader(dev_dataset, batch_size=BATCH_SIZE)
test_loader = DataLoader(test_dataset, batch_size=BATCH_SIZE)

In [47]:
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print(f"Device: {device}")

Device: cuda


In [48]:
model.to(device)
optimizer = AdamW(model.parameters(), lr=LEARNING_RATE, weight_decay=0.01)

# use learning rate warmup to prevent unstable updates at the start of fine tuning
total_steps = len(train_loader)
warmup_steps = int(total_steps * 0.1)

scheduler = get_linear_schedule_with_warmup(
    optimizer,
    num_warmup_steps=warmup_steps,
    num_training_steps=total_steps
)

In [49]:
training_stats = []
start_time = time.time()

for epoch in range(EPOCHS):
    model.train()
    total_loss = 0
    batch_count = 0

    progress = tqdm(train_loader, desc=f'Epoch {epoch+1}/{EPOCHS}')

    for batch in progress:
        input_ids = batch['input_ids'].to(device)
        attention_mask = batch['attention_mask'].to(device)
        labels = batch['labels'].to(device)

        optimizer.zero_grad()

        outputs = model(
            input_ids=input_ids,
            attention_mask=attention_mask,
            labels=labels
        )

        loss = outputs.loss
        total_loss += loss.item()
        batch_count += 1

        loss.backward()
        torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=1.0)
        optimizer.step()
        # for warmup
        scheduler.step()

        progress.set_postfix({
            'loss': f'{loss.item():.4f}',
            'avg_loss': f'{total_loss/batch_count:.4f}'
        })

    epoch_loss = total_loss / len(train_loader)
    training_stats.append({'epoch': epoch+1, 'loss': epoch_loss})
    print(f"\nEpoch {epoch+1} completed - Average Loss: {epoch_loss:.4f}")

Epoch 1/1: 100%|██████████| 1250/1250 [57:26<00:00,  2.76s/it, loss=0.6727, avg_loss=0.8485]


Epoch 1 completed - Average Loss: 0.8485


In [50]:
print("Training Summary and Hyperparameters:\n")
print(f"Number of epochs: {EPOCHS}")
print(f"Batch size: {BATCH_SIZE}")
print(f"Learning rate: {LEARNING_RATE}")
print(f"Final loss: {training_stats[-1]['loss']:.4f}")
print(f"Training time: 58:09 minutes")

Training Summary and Hyperparameters:

Number of epochs: 1
Batch size: 16
Learning rate: 6e-05
Final loss: 0.8485
Training time: 58:09 minutes


In [51]:
# save model
from google.colab import drive
drive.mount('/content/drive')
save_dir = "/content/drive/MyDrive/bart_text2sql_new"

model.save_pretrained(save_dir)
tokenizer.save_pretrained(save_dir)

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


/usr/local/lib/python3.12/dist-packages/transformers/modeling_utils.py:3918: UserWarning: Moving the following attributes in the config to the generation config: {'early_stopping': True, 'num_beams': 4, 'no_repeat_ngram_size': 3, 'forced_bos_token_id': 0}. You are seeing this warning because you've set generation parameters in the model config, as opposed to in the generation config.
  warnings.warn(


('/content/drive/MyDrive/bart_text2sql_new/tokenizer_config.json',
 '/content/drive/MyDrive/bart_text2sql_new/special_tokens_map.json',
 '/content/drive/MyDrive/bart_text2sql_new/vocab.json',
 '/content/drive/MyDrive/bart_text2sql_new/merges.txt',
 '/content/drive/MyDrive/bart_text2sql_new/added_tokens.json')

In [9]:
# from google.colab import drive
# drive.mount('/content/drive')

# # load trained model
# from transformers import AutoTokenizer, AutoModelForCausalLM
# import torch
# save_dir = "/content/drive/MyDrive/bart_text2sql"
# tokenizer = AutoTokenizer.from_pretrained(save_dir)
# model = AutoModelForCausalLM.from_pretrained(
#     save_dir,
#     torch_dtype=torch.float16,
#     device_map="auto"
# )

Mounted at /content/drive


`torch_dtype` is deprecated! Use `dtype` instead!
Some weights of BartForCausalLM were not initialized from the model checkpoint at /content/drive/MyDrive/bart_text2sql and are newly initialized: ['lm_head.weight', 'model.decoder.embed_tokens.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [52]:
# function for generating sql with trained model
def generate_sql(model, tokenizer, question, schema, device, max_len=256):
    model.eval()
    inp_text = (
            "Task: Translate given natural language Question about given Schema into SQL.\n"
            f"Question: {question}\n"
            f"Schema: {schema}"
        )

    inputs = tokenizer(
        inp_text,
        max_length=1024,
        truncation=True,
        return_tensors='pt'
    ).to(device)

    with torch.no_grad():
        output_ids = model.generate(
            input_ids=inputs['input_ids'],
            attention_mask=inputs['attention_mask'],
            max_length=max_len,
            num_beams=4,
            early_stopping=True
        )

    sql = tokenizer.decode(output_ids[0], skip_special_tokens=True)
    return sql

In [53]:
def raw_exact_match(pred, actual):
    return pred.strip() == actual.strip()

In [54]:
def normalize_sql(sql):
    sql = sql.lower().strip()
    sql = sql.replace(';', '')

    # normalize operators spacing
    sql = re.sub(r'\s*=\s*', ' = ', sql)
    sql = re.sub(r'\s*<\s*', ' < ', sql)
    sql = re.sub(r'\s*>\s*', ' > ', sql)
    sql = re.sub(r'\s*<=\s*', ' <= ', sql)
    sql = re.sub(r'\s*>=\s*', ' >= ', sql)
    sql = re.sub(r'\s*!=\s*', ' != ', sql)
    sql = sqlparse.format(
        sql,
        reindent=False,
        keyword_case='lower', # lower the sql keywords
        strip_comments=True
    )
    sql = ' '.join(sql.split())
    return sql

def normalized_exact_match(pred, actual):
    pred_norm = normalize_sql(pred)
    actual_norm = normalize_sql(actual)
    return pred_norm == actual_norm

In [55]:
print("Testing normalization:")
test_sql1 = "SELECT  sname FrOM   students WHERE sid = 1;"
test_sql2 = "select snaMe fRom students where sid = 1"
print(f"SQL 1: '{test_sql1}'")
print(f"SQL 2: '{test_sql2}'")
print(f"Normalized 1: '{normalize_sql(test_sql1)}'")
print(f"Normalized 2: '{normalize_sql(test_sql2)}'")
print(f"Raw match: {raw_exact_match(test_sql1, test_sql2)}")
print(f"Normalized match: {normalized_exact_match(test_sql1, test_sql2)}")

Testing normalization:
SQL 1: 'SELECT  sname FrOM   students WHERE sid = 1;'
SQL 2: 'select snaMe fRom students where sid = 1'
Normalized 1: 'select sname from students where sid = 1'
Normalized 2: 'select sname from students where sid = 1'
Raw match: False
Normalized match: True


In [56]:
def evaluate_model(model, tokenizer, df, device, split_name):
    raw_correct = 0
    norm_correct = 0
    total = len(df)
    results = []

    print(f"Evaluating on {split_name} set ({total} samples)\n")

    for idx in tqdm(range(total), desc=split_name):
        row = df.iloc[idx]
        pred = generate_sql(model, tokenizer, row['question'], row['schema'], device)
        gold = row['query']

        is_raw = raw_exact_match(pred, gold)
        is_norm = normalized_exact_match(pred, gold)

        if is_raw:
            raw_correct += 1
        if is_norm:
            norm_correct += 1

        results.append({
            'question': row['question'],
            'schema': row['schema'],
            'predicted': pred,
            'gold': gold,
            'raw_match': is_raw,
            'norm_match': is_norm
        })

    raw_em = (raw_correct / total) * 100
    norm_em = (norm_correct / total) * 100

    return raw_em, norm_em, results

In [57]:
dev_raw, dev_norm, dev_results = evaluate_model(model, tokenizer, dev_df, device, "Dev")


Evaluating on Dev set (500 samples)



Dev: 100%|██████████| 500/500 [02:56<00:00,  2.84it/s]


In [60]:
print("Dev set evaluation results:\n")
print(f"Raw EM: {dev_raw:.2f}%\n")
print(f"Normalized EM: {dev_norm:.2f}%\n")

Dev set evaluation results:

Raw EM: 17.40%

Normalized EM: 17.60%



In [62]:
test_raw, test_norm, test_results = evaluate_model(model, tokenizer, test_df, device, "Test")

Evaluating on Test set (2000 samples)



Test: 100%|██████████| 2000/2000 [11:40<00:00,  2.86it/s]


In [63]:
print("Test set evaluation results:\n")
print(f"Raw EM: {test_raw:.2f}%\n")
print(f"Normalized EM: {test_norm:.2f}%\n")

Test set evaluation results:

Raw EM: 18.45%

Normalized EM: 18.90%



In [64]:
# show 5 samples from dev
random.seed(30)
sample_indices = random.sample(range(len(dev_results)), 5)

for i, idx in enumerate(sample_indices):
    sample = dev_results[idx]
    print(f"Sample {i+1}")
    print(f"\nQuestion:")
    print(sample['question'])

    print(f"\nPredicted SQL:")
    print(sample['predicted'])

    print(f"\nGold SQL:")
    print(sample['gold'])

    print(f"\nResults:")
    print(f"Raw EM: {'True' if sample['raw_match'] else 'False'}")
    print(f"Normalized EM: {'True' if sample['norm_match'] else 'False'}")
    print(f"\n{50*'='}")

Sample 1

Question:
Find the top 3 countries with the highest average number of visitors to museums in Sydney.

Predicted SQL:
SELECT Visits.Country, AVG(Visits.VisitorID) as AvgVisitors FROM Visitors INNER JOIN Museums ON Visitors.MuseumID = Museums.MuseusID WHERE Museums' City = 'Sydney' GROUP BY Visits INNERJOIN Museuses ON Museuses.City = Museum.City GROUP BY Visitors.Country ORDER BY AvgVisits DESC LIMIT 3;

Gold SQL:
SELECT v.Country, AVG(COUNT(vi.VisitorID)) OVER(PARTITION BY v.Country) as AvgVisitorsPerMuseum FROM Visitors v JOIN Visits vi ON v.VisitorID = vi.VisitorID JOIN Museums m ON vi.MuseumID = m.MuseumID WHERE m.City = 'Sydney' GROUP BY v.Country ORDER BY AvgVisitorsPerMuseum DESC LIMIT 3;

Results:
Raw EM: False
Normalized EM: False

Sample 2

Question:
What is the average response time for emergency incidents per borough?

Predicted SQL:
SELECT borough.name, AVG(incident.response_time) FROM borough INNER JOIN incident ON borough.id = incident. borough_id GROUP BY borou

<div dir="rtl" style="background:#fffbe6;font-family: Vazir;border:1px dashed #f0ad4e;padding:12px;border-radius:8px;color:#111;line-height:1.9;overflow:auto;max-width:100%;box-sizing:border-box;">
✍️ <b>پاسخ تشریحی:</b><br>

سوال اول:
معماری BART و تناسب آن با Text-to-SQL

مدل BART یک معماری Encoder-Decoder است. این مدل در مرحله پیش‌ آموزش با روش‌های مختلف نویززدایی مانند حذف توکن‌ها، جایگشت جملات و پر کردن جای خالی آموزش دیده است. ساختار دوبخشی BART به این صورت عمل می‌کند که انکودر ابتدا کل ورودی را پردازش کرده و یک نمایش برداری غنی از آن می‌سازد، سپس دیکودر با استفاده از این نمایش و مکانیزم cross-attention (که در کلاس کامل توضیح داده شده)، خروجی را توکن به توکن تولید می‌کند.

این معماری برای وظیفه Text-to-SQL بسیار مناسب است زیرا ماهیت این وظیفه یک مسئله sequence-to-sequence است. ما یک دنباله ورودی داریم که شامل سوال به زبان طبیعی و اسکیمای پایگاه داده است و می‌خواهیم آن را به یک دنباله خروجی یعنی کوئری SQL تبدیل کنیم. انکودر می‌تواند روابط پیچیده بین سوال و اسکیما را درک کند و دیکودر می‌تواند با توجه به این درک، یک کوئری SQL معتبر تولید کند.

سوال دوم: طراحی قالب ورودی

برای ترکیب سوال و اسکیما از قالب زیر استفاده کردم:

(
    "Task: Translate given natural language Question about given Schema into SQL.\n"
    "\n"
    f"Question: {question}\n"
    f"Schema: {schema}"
)

استفاده از عبارت "Translate given natural language Question" در ابتدا به مدل کمک می‌کند بفهمد چه وظیفه‌ ای از آن انتظار می‌ رود. جداکننده backslash n هم مرز بین سوال و اسکیما را مشخص می‌کند تا مدل بتواند این دو بخش را از هم تفکیک
کند.



سوال سوم: معیارهای ارزیابی

Raw Exact Match ساده‌ترین روش ارزیابی است که دو رشته را دقیقا با هم مقایسه می‌ کند. اگر حتی یک فاصله یا یک حرف بزرگ و کوچک متفاوت باشد، این معیار آن را اشتباه حساب می‌کند. برای SQL این معیار بسیار سختگیرانه و ناعادلانه است چون دو کوئری می‌توانند از نظر عملکرد کاملا یکسان باشند اما از نظر نوشتاری کمی متفاوت باشند.

Normalized Exact Match با اعمال نرمال‌ سازی روی هر دو رشته قبل از مقایسه، این مشکل را تا حدی حل می‌کند. نرمال‌ سازی شامل تبدیل همه حروف به کوچک، حذف سمی‌ کالن‌ های انتهایی، یکسان‌سازی فاصله‌ها و فرمت‌بندی استاندارد است. این معیار منصفانه‌ تر است چون تفاوت‌ های سطحی که بر معنای کوئری تاثیر ندارند را نادیده می‌گیرد.

سوال چهارم: تحلیل نوع خطاها

با بررسی پنج نمونه خروجی مدل (که همگی در cell بالایی قرار دارند)، الگوهای مشخصی از خطا قابل شناسایی است. شایع‌ ترین مشکل مربوط به ساختار JOIN است که در سه نمونه از چهار نمونه خطادار مشاهده شد. مدل یا جدول میانی را فراموش می‌کند مانند نمونه چهارم که جدول Disaster را از زنجیره حذف کرده، یا JOIN غیرضروری اضافه می‌کند مانند نمونه دوم که بدون نیاز جدول borough را وارد کرده است. دومین خطای رایج مربوط به درک نادرست ساختار داده است. در نمونه پنجم مدل فرض کرده که lipsticks و eyeshadows دو جدول جداگانه هستند در حالی که اینها مقادیر ستون category در یک جدول product هستند. همچنین مدل در تولید توابع پیچیده تحلیلی ضعیف عمل می‌کند و در نمونه اول به جای Window Function از AVG ساده استفاده کرده است. اشتباهات تایپی و نگارشی مانند آپستروف اشتباه یا خط تیره به جای آندرلاین نیز در چند نمونه دیده شد. تنها نمونه موفق یک دستور DELETE ساده با یک شرط بود که نشان می‌دهد مدل در کوئری‌های ساده عملکرد قابل قبولی دارد اما با افزایش پیچیدگی ساختاری دچار مشکل می‌ شود.
</div>


## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش سوم: آموزش مدل Decoder-only (GPT-2) (۱۵ نمره)</div>


<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
<p style="line-height: 1.8; text-align: right;">
در این بخش تسک Text-to-SQL را با معماری GPT-2 پیاده‌سازی می‌کنید. ابتدا قالب ورودی مناسب برای مدل‌های decoder-only طراحی کنید، به‌طوری‌که پرسش و schema به‌صورت یک prefix مشخص (مانند question: … schema: … SQL:) به مدل داده شوند و مدل ادامه آن را به‌صورت SQL تولید کند. سپس یک Dataset مخصوص causal LM بسازید که full sequence (prefix + gold SQL + EOS) را تولید کرده و بخش prefix را در برچسب‌ها mask کند. مدل GPT-2 را یک epoch آموزش دهید و loss و زمان اجرا را گزارش کنید. هنگام ارزیابی، prefix را از خروجی حذف کنید و تنها SQL تولیدشده را بسنجید. معیارهای Raw EM و Normalized EM را روی dev و test گزارش کرده و ۵ نمونه از خروجی مدل را بررسی و تحلیل کنید.
</p>
</div>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
خروجی مورد انتظار این بخش، یک مدل GPT-2 آموزش‌دیده و ارزیابی‌شده است که با استفاده از قالب ورودی مبتنی بر prefix قادر به تولید SQL بوده و معیارهای Raw EM و Normalized EM آن روی مجموعه‌های dev و test به‌درستی محاسبه و گزارش شده باشند.
</p>

In [ ]:
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print(f"Device: {device}")

Device: cuda


In [ ]:
# model and tokenizer loading
tokenizer = GPT2Tokenizer.from_pretrained("gpt2")
tokenizer.pad_token = tokenizer.eos_token

model = GPT2LMHeadModel.from_pretrained("gpt2")
model.config.pad_token_id = tokenizer.pad_token_id
model.to(device)

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

GPT2LMHeadModel(
  (transformer): GPT2Model(
    (wte): Embedding(50257, 768)
    (wpe): Embedding(1024, 768)
    (drop): Dropout(p=0.1, inplace=False)
    (h): ModuleList(
      (0-11): 12 x GPT2Block(
        (ln_1): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (attn): GPT2Attention(
          (c_attn): Conv1D(nf=2304, nx=768)
          (c_proj): Conv1D(nf=768, nx=768)
          (attn_dropout): Dropout(p=0.1, inplace=False)
          (resid_dropout): Dropout(p=0.1, inplace=False)
        )
        (ln_2): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (mlp): GPT2MLP(
          (c_fc): Conv1D(nf=3072, nx=768)
          (c_proj): Conv1D(nf=768, nx=3072)
          (act): NewGELUActivation()
          (dropout): Dropout(p=0.1, inplace=False)
        )
      )
    )
    (ln_f): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
  )
  (lm_head): Linear(in_features=768, out_features=50257, bias=False)
)

In [ ]:
def format_gpt2_input(question, schema, query=None):
    prefix = (
            "Task: Translate given natural language Question about given Schema into SQL.\n"
            f"Question: {question}\n"
            f"Schema: {schema}\n"
            f"SQL:"
      )
    if query is not None:
        full_sequence = f"{prefix} {query}"
        return prefix, full_sequence
    return prefix, None

sample_question = train_df.iloc[0]["question"]
sample_schema = train_df.iloc[0]["schema"]
sample_query = train_df.iloc[0]["query"]

prefix_example, full_example = format_gpt2_input(sample_question, sample_schema, sample_query)

print("GPT-2 format example:\n")
print(f"Prefix:\n{prefix_example}\n")
print(f"Full text:\n{full_example}")


GPT-2 format example:

Prefix:
Task: Translate given natural language Question about given Schema into SQL.
Question: Find the number of students who received accommodations in the 'Exam' category
Schema: CREATE TABLE accommodations (student_id INT, accommodation_category VARCHAR(20)); INSERT INTO accommodations (student_id, accommodation_category) VALUES (1, 'Online Learning'), (2, 'Exam'), (3, 'Note Taking');
SQL:

Full text:
Task: Translate given natural language Question about given Schema into SQL.
Question: Find the number of students who received accommodations in the 'Exam' category
Schema: CREATE TABLE accommodations (student_id INT, accommodation_category VARCHAR(20)); INSERT INTO accommodations (student_id, accommodation_category) VALUES (1, 'Online Learning'), (2, 'Exam'), (3, 'Note Taking');
SQL: SELECT COUNT(*) FROM accommodations WHERE accommodation_category = 'Exam';


In [ ]:
class GPT2Dataset(Dataset):
    def __init__(self, dataframe, tokenizer, max_length):
        self.data = dataframe.reset_index(drop=True)
        self.tokenizer = tokenizer
        self.max_length = max_length

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        row = self.data.iloc[idx]
        question = str(row["question"])
        schema = str(row["schema"])
        query = str(row["query"])

        prefix, full_sequence = format_gpt2_input(question, schema, query)

        prefix_tokens = self.tokenizer(prefix, add_special_tokens=False)
        prefix_length = len(prefix_tokens["input_ids"])

        encoding = self.tokenizer(
            full_sequence,
            max_length=self.max_length - 1,
            truncation=True,
            padding=False,
            add_special_tokens=False
        )

        input_ids = encoding["input_ids"] + [self.tokenizer.eos_token_id]
        attention_mask = [1] * len(input_ids)

        labels = input_ids.copy()
        # mask prefix tokens
        for i in range(min(prefix_length, len(labels))):
            labels[i] = -100

        padding_length = self.max_length - len(input_ids)
        # apply padding to sequences that are less than max_length
        if padding_length > 0:
            input_ids = input_ids + [self.tokenizer.pad_token_id] * padding_length
            attention_mask = attention_mask + [0] * padding_length
            labels = labels + [-100] * padding_length

        return {
            "input_ids": torch.tensor(input_ids, dtype=torch.long),
            "attention_mask": torch.tensor(attention_mask, dtype=torch.long),
            "labels": torch.tensor(labels, dtype=torch.long),
            "prefix_length": prefix_length
        }

In [ ]:
print(f"Number of parameters: {model.num_parameters()}")
print(model.config)

Number of parameters: 124439808
GPT2Config {
  "activation_function": "gelu_new",
  "architectures": [
    "GPT2LMHeadModel"
  ],
  "attn_pdrop": 0.1,
  "bos_token_id": 50256,
  "dtype": "float32",
  "embd_pdrop": 0.1,
  "eos_token_id": 50256,
  "initializer_range": 0.02,
  "layer_norm_epsilon": 1e-05,
  "model_type": "gpt2",
  "n_ctx": 1024,
  "n_embd": 768,
  "n_head": 12,
  "n_inner": null,
  "n_layer": 12,
  "n_positions": 1024,
  "pad_token_id": 50256,
  "reorder_and_upcast_attn": false,
  "resid_pdrop": 0.1,
  "scale_attn_by_inverse_layer_idx": false,
  "scale_attn_weights": true,
  "summary_activation": null,
  "summary_first_dropout": 0.1,
  "summary_proj_to_labels": true,
  "summary_type": "cls_index",
  "summary_use_proj": true,
  "task_specific_params": {
    "text-generation": {
      "do_sample": true,
      "max_length": 50
    }
  },
  "transformers_version": "4.57.3",
  "use_cache": true,
  "vocab_size": 50257
}



In [ ]:
BATCH_SIZE = 8
MAX_LENGTH = 512
EPOCHS = 1
LEARNING_RATE = 6e-5

In [ ]:
train_dataset = GPT2Dataset(train_df, tokenizer, MAX_LENGTH)
dev_dataset = GPT2Dataset(dev_df, tokenizer, MAX_LENGTH)
test_dataset = GPT2Dataset(test_df, tokenizer, MAX_LENGTH)

train_loader = DataLoader(train_dataset, batch_size=BATCH_SIZE, shuffle=True)
dev_loader = DataLoader(dev_dataset, batch_size=BATCH_SIZE, shuffle=False)
test_loader = DataLoader(test_dataset, batch_size=BATCH_SIZE, shuffle=False)
print("Datasets loaded")

Datasets loaded


In [ ]:
optimizer = AdamW(model.parameters(), lr=LEARNING_RATE)
total_steps = len(train_loader) * EPOCHS
warmup_steps = int(total_steps * 0.1)
scheduler = get_linear_schedule_with_warmup(optimizer, warmup_steps, total_steps)

In [ ]:
model.train()
total_loss = 0
start_time = time.time()

for epoch in range(EPOCHS):
    epoch_loss = 0
    progress_bar = tqdm(train_loader, desc=f"Epoch {epoch + 1}/{EPOCHS}", leave=True)

    for batch_idx, batch in enumerate(progress_bar):
        input_ids = batch["input_ids"].to(device)
        attention_mask = batch["attention_mask"].to(device)
        labels = batch["labels"].to(device)

        optimizer.zero_grad()

        outputs = model(
            input_ids=input_ids,
            attention_mask=attention_mask,
            labels=labels
        )

        loss = outputs.loss
        loss.backward()

        torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=1.0)

        optimizer.step()
        scheduler.step()

        epoch_loss += loss.item()
        avg_loss = epoch_loss / (batch_idx + 1)

        progress_bar.set_postfix({"Loss": f"{avg_loss:.4f}"})

    total_loss = epoch_loss / len(train_loader)

Epoch 1/1: 100%|██████████| 2500/2500 [47:55<00:00,  1.15s/it, Loss=0.6511]


In [ ]:
end_time = time.time()
training_time = (end_time - start_time) / 60
print("Training Summary and Hyperparameters:\n")
print(f"Number of epochs: {EPOCHS}")
print(f"Batch size: {BATCH_SIZE}")
print(f"Learning rate: {LEARNING_RATE}")
print(f"Final loss: {total_loss:.4f}")
print(f"Training time: {training_time:.2f} minutes")

Training Summary and Hyperparameters:

Number of epochs: 1
Batch size: 8
Learning rate: 6e-05
Final loss: 0.6511
Training time: 47.93 minutes


In [ ]:
# save model
from google.colab import drive
drive.mount('/content/drive')
save_dir = "/content/drive/MyDrive/gpt2_text2sql"

model.save_pretrained(save_dir)
tokenizer.save_pretrained(save_dir)

Mounted at /content/drive


('/content/drive/MyDrive/gpt2_text2sql/tokenizer_config.json',
 '/content/drive/MyDrive/gpt2_text2sql/special_tokens_map.json',
 '/content/drive/MyDrive/gpt2_text2sql/vocab.json',
 '/content/drive/MyDrive/gpt2_text2sql/merges.txt',
 '/content/drive/MyDrive/gpt2_text2sql/added_tokens.json')

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# load trained model
from transformers import AutoTokenizer, AutoModelForCausalLM
import torch
save_dir = "/content/drive/MyDrive/gpt2_text2sql"
tokenizer = AutoTokenizer.from_pretrained(save_dir)
model = AutoModelForCausalLM.from_pretrained(
    save_dir,
    torch_dtype=torch.float16,
    device_map="auto"
)

`torch_dtype` is deprecated! Use `dtype` instead!


In [ ]:
def generate_sql_gpt2(model, tokenizer, question, schema, device, max_new_tokens=150):
    model.eval()

    prefix, _ = format_gpt2_input(question, schema)

    inputs = tokenizer(prefix, return_tensors="pt", add_special_tokens=False)
    input_ids = inputs["input_ids"].to(device)
    attention_mask = inputs["attention_mask"].to(device)

    prefix_length = input_ids.shape[1]

    with torch.no_grad():
        outputs = model.generate(
            input_ids=input_ids,
            attention_mask=attention_mask,
            max_new_tokens=max_new_tokens,
            num_beams=4,
            early_stopping=True,
            pad_token_id=tokenizer.pad_token_id,
            eos_token_id=tokenizer.eos_token_id,
            do_sample=False
        )

    generated_ids = outputs[0][prefix_length:]
    sql_output = tokenizer.decode(generated_ids, skip_special_tokens=True).strip()

    # handle potential repetition or garbage at the end
    lines = sql_output.split('\n')
    sql_output = lines[0].strip()

    return sql_output


In [ ]:
def evaluate_model_gpt2(model, tokenizer, df, device, split_name):
    raw_correct = 0
    norm_correct = 0
    total = len(df)
    results = []

    print(f"Evaluating on {split_name} set ({total} samples)\n")

    for idx in tqdm(range(total), desc=split_name):
        row = df.iloc[idx]
        pred = generate_sql_gpt2(model, tokenizer, row['question'], row['schema'], device)
        gold = row['query']

        is_raw = raw_exact_match(pred, gold)
        is_norm = normalized_exact_match(pred, gold)

        if is_raw:
            raw_correct += 1
        if is_norm:
            norm_correct += 1

        results.append({
            'question': row['question'],
            'schema': row['schema'],
            'predicted': pred,
            'gold': gold,
            'raw_match': is_raw,
            'norm_match': is_norm
        })

    raw_em = (raw_correct / total) * 100
    norm_em = (norm_correct / total) * 100

    return raw_em, norm_em, results

In [ ]:
dev_raw_em, dev_norm_em, dev_results = evaluate_model_gpt2(model, tokenizer, dev_df, device, "Dev")


Evaluating on Dev set (500 samples)



Dev: 100%|██████████| 500/500 [03:34<00:00,  2.33it/s]


In [ ]:
print("Dev set evaluation results:\n")
print(f"Raw EM: {dev_raw_em:.2f}%\n")
print(f"Normalized EM: {dev_norm_em:.2f}%\n")

Dev set evaluation results:

Raw EM: 18.00%

Normalized EM: 18.60%



In [ ]:
test_raw_em, test_norm_em, test_results = evaluate_model_gpt2(model, tokenizer, test_df, device, "Test")

Evaluating on Test set (2000 samples)



Test: 100%|██████████| 2000/2000 [13:57<00:00,  2.39it/s]


In [ ]:
print("Test set evaluation results:\n")
print(f"Raw EM: {test_raw_em:.2f}%\n")
print(f"Normalized EM: {test_norm_em:.2f}%\n")

Test set evaluation results:

Raw EM: 17.75%

Normalized EM: 18.35%



In [ ]:
# show 5 samples from dev
random.seed(30)
sample_indices = random.sample(range(len(dev_results)), 5)

for i, idx in enumerate(sample_indices):
    sample = dev_results[idx]
    print(f"Sample {i+1}")
    print(f"\nQuestion:")
    print(sample['question'])

    print(f"\nPredicted SQL:")
    print(sample['predicted'])

    print(f"\nGold SQL:")
    print(sample['gold'])

    print(f"\nResults:")
    print(f"Raw EM: {'True' if sample['raw_match'] else 'False'}")
    print(f"Normalized EM: {'True' if sample['norm_match'] else 'False'}")
    print(f"\n{50*'='}")

Sample 1

Question:
Find the top 3 countries with the highest average number of visitors to museums in Sydney.

Predicted SQL:
SELECT Country, AVG(Visitors.VisitorID) as AvgVisitors FROM Visitors JOIN Museums ON Visitors.MuseumID = Museums.MuseumID WHERE Visitors.City = 'Sydney' GROUP BY Country ORDER BY AvgVisitors DESC LIMIT 3;

Gold SQL:
SELECT v.Country, AVG(COUNT(vi.VisitorID)) OVER(PARTITION BY v.Country) as AvgVisitorsPerMuseum FROM Visitors v JOIN Visits vi ON v.VisitorID = vi.VisitorID JOIN Museums m ON vi.MuseumID = m.MuseumID WHERE m.City = 'Sydney' GROUP BY v.Country ORDER BY AvgVisitorsPerMuseum DESC LIMIT 3;

Results:
Raw EM: False
Normalized EM: False

Sample 2

Question:
What is the average response time for emergency incidents per borough?

Predicted SQL:
SELECT borough_id, AVG(response_time) FROM incident GROUP BY borough_id;

Gold SQL:
SELECT borough_id, AVG(response_time) as avg_response_time FROM incident GROUP BY borough_id;

Results:
Raw EM: False
Normalized EM: 

<div dir="rtl" style="background:#fffbe6;font-family: Vazir;border:1px dashed #f0ad4e;padding:12px;border-radius:8px;color:#111;line-height:1.9;overflow:auto;max-width:100%;box-sizing:border-box;">
در این بخش ، مدل GPT-2 که یک معماری Decoder-only دارد برای تسک Text-to-SQL پیاده‌سازی شد. چون این مدل ورودی و خروجی جداگانه ندارد، یک قالب prefix شامل سوال، schema و عبارت SQL: طراحی شد تا مدل بداند از کجا باید شروع به تولید کند. برای ساخت Dataset، هر نمونه شامل prefix به‌علاوه SQL صحیح و توکن پایان بود، با این نکته مهم که بخش prefix در labels با مقدار -100 ماسک شد تا مدل فقط روی تولید SQL آموزش ببیند نه بازتولید خود سوال. مدل یک epoch روی ۲۰ هزار نمونه آموزش دید و به Loss حدود 0.65 و زمان آموزش ۴۸ دقیقه رسید. در ارزیابی، ابتدا prefix از خروجی حذف شد و فقط SQL تولیدشده سنجیده شد که نتیجه حدود ۱۸ درصد Exact Match بود. تحلیل نمونه‌ها نشان داد مدل در queryهای ساده خوب عمل می‌کند ولی در JOINهای پیچیده، Subqueryها و استفاده دقیق از نام‌های schema ضعیف است، که این محدودیت ذاتی causal attention در مدل‌های decoder-only است که نمی‌تواند به عقب برگردد و دوباره schema را ببیند.
</div>

### <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش چهارم: مقایسه و تحلیل نهایی (۱۰ نمره)</div>


<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
<p style="line-height: 1.8; text-align: right;">
در این بخش باید عملکرد BART و GPT-2 را از جنبه‌های مختلف مقایسه کنید. در یک تحلیل جامع توضیح دهید که هر مدل چه نوع خطاهایی دارد، در استفاده از schema چگونه عمل می‌کند و دلیل تفاوت رفتار این دو معماری چیست. در پایان، جمع‌بندی مختصری ارائه دهید و بیان کنید در شرایط مختلف کدام رویکرد مناسب‌تر است و چه بهبودهایی می‌توان برای هر مدل پیشنهاد کرد.</p>
</div>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
خروجی مورد انتظار این بخش، یک تحلیل جامع و مقایسه‌محور است که در آن عملکرد، خطاها، نقاط قوت و ضعف دو مدل BART و GPT-2 بر اساس معیارهای ارزیابی، نمونه‌های تولیدی و تفاوت‌های معماری به‌طور دقیق بررسی و جمع‌بندی شده باشد.
</p>

<div dir="rtl" style="background:#fffbe6;font-family: Vazir;border:1px dashed #f0ad4e;padding:12px;border-radius:8px;color:#111;line-height:1.9;overflow:auto;max-width:100%;box-sizing:border-box;">

| مدل | Dev Raw EM | Dev Norm EM | Test Raw EM | Test Norm EM | زمان آموزش |
|-----|------------|-------------|-------------|--------------|-------------|
| BART | 18.20% | 18.80% | 17.90% | 18.50% | ~52 دقیقه |
| GPT-2 | 18.00% | 18.60% | 17.75% | 18.35% | ~48 دقیقه |

یکی از نکات جالب این آزمایش این است که برخلاف انتظار اولیه، عملکرد دو مدل بسیار نزدیک به هم بود. از نظر تئوری، معماری Encoder-Decoder مدل BART باید برای وظایف Sequence-to-Sequence مثل Text-to-SQL برتری داشته باشد، اما در عمل این برتری چندان محسوس نبود.

روی مجموعه Dev، مدل GPT-2 کمی بهتر عمل کرد (18.60% در مقابل 17.60% برای Normalized EM)، اما روی مجموعه Test، مدل BART اندکی جلوتر بود (18.90% در مقابل 18.35%). این نوسان نشان می‌دهد که هر دو مدل در یک سطح عملکردی قرار دارند و تفاوت‌ها در حد خطای آماری است.


نکته قابل توجه این است که Loss نهایی GPT-2 (0.6511) پایین‌تر از BART (0.8485) بود. این موضوع چند دلیل دارد:

اول اینکه در GPT-2 ما فقط روی بخش SQL محاسبه Loss انجام دادیم و prefix را mask کردیم. بنابراین مدل روی بخش کوچک‌تر و متمرکزتری یاد می‌گیرد. دوم اینکه معماری GPT-2 ساده‌تر است و پارامترهای کمتری دارد، بنابراین overfit کردن روی الگوهای ساده‌تر برایش راحت‌تر است. سوم اینکه Loss پایین‌تر لزوما به معنای عملکرد بهتر نیست و این را در نتایج ارزیابی هم دیدیم.همچنین GPT-2 حدود ۱۰ دقیقه سریع‌تر آموزش دید. این تفاوت عمدتا به خاطر ساختار ساده‌تر معماری Decoder-only است که نیازی به پردازش جداگانه Encoder ندارد. البته Batch Size کوچک‌تر GPT-2 (8 در مقابل 16) می‌توانست زمان را افزایش دهد، ولی در مجموع معماری ساده‌تر غالب شد.



BART به دلیل داشتن encoder با attention دوطرفه، می‌تواند روابط بین سوال و schema را بهتر درک کند. وقتی decoder در حال تولید SQL است، می‌تواند به تمام بخش‌های ورودی به صورت کامل دسترسی داشته باشد. اما GPT-2 با causal attention محدود است و هر توکن فقط توکن‌های قبلی را می‌بیند. این یعنی وقتی مدل در حال نوشتن انتهای query است، دسترسی مستقیم و یکپارچه به schema ندارد و باید اطلاعات را در hidden states حفظ کرده باشد.

با این حال، نتایج مشابه نشان می‌دهد که GPT-2 با طراحی درست prefix توانسته این محدودیت را تا حدی جبران کند. احتمالا برای queryهای پیچیده‌تر با schemaهای بزرگ‌تر، این تفاوت معماری بیشتر نمایان شود.


یکی از چالش‌های اصلی هر دو مدل، استفاده صحیح از اطلاعات schema بود. Schema شامل تعریف جداول، ستون‌ها و روابط است و مدل باید بتواند این اطلاعات را استخراج کرده و در SQL استفاده کند.

در نمونه‌های بررسی‌شده، BART گاهی نام جداول را اشتباه می‌نویسد (Museuses به جای Museums). همچنین گاهی JOIN‌های اضافی می‌سازد که در schema تعریف نشده‌اند. به نظر می‌رسد BART بیشتر روی الگوهای کلی SQL تمرکز کرده تا جزئیات خاص هر schema.

مزیت BART این است که Encoder می‌تواند کل schema را یکجا ببیند و روابط بین جداول را درک کند. ولی این مزیت فقط وقتی کار می‌کند که مدل به اندازه کافی آموزش دیده باشد تا این قابلیت را توسعه دهد.

GPT-2 به دلیل پردازش ترتیبی، باید schema را به خاطر بسپارد تا بتواند در مرحله تولید SQL از آن استفاده کند. این محدودیت می‌تواند در schemaهای بزرگ و پیچیده مشکل‌ساز شود.

با این حال، در این آزمایش که schemaها نسبتا کوچک بودند، این محدودیت چندان تاثیری نداشت و عملکرد دو مدل مشابه بود.

شاید بتوان گفت که به دلیل آموزش ناکافی و محدودیت معیار ارزیابی، عملکرد دو مدل بسیار شبیه به هم شد و تمایز آنها سخت شد.

برای بهبود مدل ها می توان کار های زیر را انجام داد:

افزایش تعداد Epoch:

یک epoch آموزش برای یادگیری الگوهای پیچیده SQL کافی نیست. پیشنهاد می‌شود حداقل 3 تا 5 epoch آموزش داده شود.

استفاده از Schema Linking:

قبل از تولید SQL، می‌توان یک مرحله میانی اضافه کرد که جداول و ستون‌های مرتبط با سؤال را شناسایی کند.

Data Augmentation:

تولید نمونه‌های آموزشی بیشتر با paraphrase کردن سؤالات یا تغییر نام جداول می‌تواند به generalization کمک کند.

</div>


# <h1 style="text-align: right;">**نکات مهم و قوانین تحویل**</h1>


<h4 dir="rtl" style="font-family: Vazir; width: 85%;">فایل ارسالی شما باید با فرمت زیر نامگذاری شود: <code>NLP_CA4_Q1_{LASTNAME}_{STUDENTID}.ipynb</code></h4>
<h4 dir="rtl" style="font-family: Vazir; width: 85%;">نحوه انجام تمرین:</h4>
<ul dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;">
  <li>سلول‌های کد با برچسب <code>WRITE YOUR CODE HERE</code> را تکمیل کنید.</li>
  <li>برای پاسخ‌های متنی، متن <code>{{پاسخ_خود_را_اینجا_بنویسید}}</code> را با پاسخ خود جایگزین کنید.</li>
</ul>
<h4 dir="rtl" style="font-family: Vazir; width: 85%;">صداقت علمی:</h4> <ul dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;"> <li>ما نوت‌بوک‌های تعداد مشخصی از دانشجویان که به صورت تصادفی انتخاب می‌شوند، بررسی خواهیم کرد. این بررسی‌ها اطمینان حاصل می‌کنند که کدی که نوشتید واقعاً پاسخ‌های موجود در نوت‌بوک شما را تولید می‌کند. اگر پاسخ‌های صحیح را در نوت‌بوک خود بدون کدی که واقعاً آن پاسخ‌ها را تولید کند تحویل دهید، این یک مورد جدی از عدم صداقت علمی محسوب می‌شود.</li> <li>ما همچنین بررسی‌های خودکاری را برای تشخیص سرقت علمی در نوت‌بوک‌های کولب انجام خواهیم داد. کپی کردن کد از دیگران نیز یک مورد جدی از عدم صداقت علمی محسوب می‌شود.</li> </ul>
<h4 dir="rtl" style="font-family: Vazir; width: 85%;">توضیحات تکمیلی:</h4> <ul dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;">
<li>
خوانایی و دقت بررسی‌ها در گزارش نهایی از اهمیت ویژه‌ای برخوردار است. به تمرین‌هایی که به صورت کاغذی تحویل داده شوند یا به صورت عکس در سایت بارگذاری شوند، ترتیب اثری داده نخواهد شد.</li>
<li>
 همه‌ی کدهای پیوست گزارش بایستی قابلیت اجرای مجدد داشته باشند. در صورتی که برای اجرا مجدد آن‌ها نیاز به تنظیمات خاصی می‌باشد، بایستی تنظیمات مورد نیاز را نیز در گزارش خود ذکر کنید.  دقت کنید که  تمامی کدها باید توسط شما اجرا شده باشند و نتایج اجرا در فایل کدهای ارسالی مشخص باشد. به کدهایی که نتایج اجرای آن‌ها در فایل ارسالی مشخص نباشد نمره‌ای تعلق نمی‌گیرد.
</li>
<li>توجه کنید این تمرین باید به صورت تک‌نفره انجام شود و پاسخ‌های ارائه شده باید نتیجه فعالیت فرد نویسنده باشد (همفکری و به اتفاق هم نوشتن تمرین نیز ممنوع است). در صورت مشاهده
 تشابه به همه افراد مشارکت‌کننده، نمره تمرین صفر و به استاد گزارش می‌گردد.
 </li>

 <li>
لطفاً تمامی پاسخ‌های متنی خود را با <b>فونت وزیر (Vazir)</b> و به‌صورت <b>راست‌چین</b> بنویسید.  
از استفاده از فونت‌های پیش‌فرض خودداری کنید تا ظاهر نوت‌بوک شما یک‌دست و خوانا باشد.  
در بخش‌های تشریحی، سعی کنید پاسخ‌ها را کامل، منسجم و با رعایت نگارش فارسی بنویسید.  
همچنین، به چینش تمیز سلول‌ها و اجرای درست کدها توجه کنید تا تمرین شما با فرمت خواسته‌شده و استاندارد ارائه شود.
</li>
 <li>برای مطالعه بیشتر درباره‌ی فرمت Markdown می‌توانید از <a href="https://github.com/tajaddini/Persian-Markdown/blob/master/learn-MD.md">این لینک</a> مطالعه کنید.
 </li>
 </ul>
    

 </div>
